### ISSUE TITLE
Excessive instantiation depth with fluent chaining API

### ISSUE THREAD

**Issue by andrewbranch** [microsoft/TypeScript#46783]
@amcasey reduced a perf complaint from https://github.com/glideapps/glide-code-columns/blob/master/src/columns/as-array.ts to [this playground](https://www.typescriptlang.org/play?ts=4.6.0-dev.20211105#code/MYGwhgzhAEDCD2IA8AVACmATmAthAfNAN4BQ050ADgK4BGIAlsNAO4MAuAFgEoCmAjtQaZeAEzSYGODgwBuvDNhyoAcrl7ReAD3a8AdqJgR2kvQHN8ACj3qAXNBRqcvAJTEyFTyPbVMe6NQGvABmDHpiAITQkHCISB6eiQ6KuDAAZMTQImCi8HogAJ7QANoA0tBhDk68ALr2YHpFAL4JifgA3K0tLSThLLEgli6tAHRsXHyCwmISUjLyKTiWAESyYCDUvMvDnmMcPAJCIuKS0uxyCli4K2sbWzsUexOH0ydz5wtXS6vrm9uj4wOU2OszOF0WN1+9wB+0mRxmp3mlyUkLu-12gLhr1BSIhPzRD3ITyB8LeYM+KPxf0J0GJWJBiI+yOuVOhGNhLwZ73BX1R1Jhz2BCO5FJZt357MFpJxTLx4rZj0xnOF5OZ33l6MVHKFZNxvNZmqJSp1Mp5lI1NLpyt1sv1FugAHoHdAABIcIzsMDAADW0EYZyAA):

```ts
class Col<TParams> {
    public withRequiredPrimitiveParam<TName extends string>(name: TName) {
        return undefined! as Col<
            TParams & { readonly [K in TName]: any }
        >;
    }
}

new Col()
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value")
    .withRequiredPrimitiveParam("value") // Hits stack limit
```

Each additional chain results in a new instantiation of `Col` with an additional `{ value: any }` intersected with the previous instantiation’s type argument. These intersections don’t seem to get simplified for some reason—hovering the third call in the chain yields the quick info

```ts
(method) Col<{ readonly value: any; } & { readonly value: any; }>.withRequiredPrimitiveParam<"value">(name: "value"): Col<{
    readonly value: any;
} & {
    readonly value: any;
} & {
    readonly value: any;
}>
```

Working with a more realistic simplification of the source, @amcasey measured type instantiations as a function of calls in the chain:

![Chart showing exponential increase of instantiations, skyrocketing at around 12 calls](https://user-images.githubusercontent.com/3277153/141338167-39adbfaf-9cd4-4a5d-b742-f6710d97a8e6.png)

It seems that the function is the same even for this dead simple reproduction where each call after the first doesn’t meaningfully change the return type since we max out at the same number of calls, and it’s not obvious to me why that pattern is exponential.

**Comment by amcasey** [comment]
In the original repro, you only need `src/columns/as-array.ts` (with duplicated calls) and `src/glide.ts`.  Dropping the declarations supported by imports in `glide.ts` has no effect.

**Comment by ahejlsberg** [comment]
First, the easy workaround:

```ts
class Col<TParams> {
    public withRequiredPrimitiveParam<TName extends string>(name: TName) {
        return undefined! as Col<
            TParams & Record<TName, any>
        >;
    }
}
```

Basically, replace `{ readonly [K in TName]: any }` with `Record<TName, any>`. What's the difference? Well, in order to maximally reuse instantiations of anonymous object types we determine which in-scope type parameters are possibly referenced within the object type and then create a map keyed by the type IDs for which the type parameters are instantiated. The object type only references `TName`, so that should be the only type parameter contributing to the key. However, it is hard for us to prove that `TParams` _isn't_ referenced because there is intervening non-type-declaration code between the object type and the declaration of `TParams`. For example, there could be a local type that references `TParams` and a reference to that local type in the anonymous object type. For that reason, we view `TParams` as possibly referenced and include it in the instantiation key map. That in turn causes all the instantiations to look different to us, so we keep manufacturing new ones, making more and more types until we eventually hit our governor.

If we instead write `Record<TName, any>`, we effectively remove `TParams` from scope and all the instantiations collapse to a single cached type.

I'll think about ways we can be smarter about concluding that type parameters aren't referenced, but it's not trivial.

**Comment by fatcerberus** [comment]
Doesn't `Record<TName, any>` expand to `{ [K in TName]: any }`, though? Since `Record` is defined as a type alias (AFAIK), equational reasoning seems to suggest the same limitations should apply...

**Comment by ahejlsberg** [comment]
> Doesn't `Record<TName, any>` expand to `{ [K in TName]: any }`, though?

Yes, they are structurally identical. But they are still represented as two distinct type instances in the type checker (specifically, two distinct instances of the `ObjectType` type), and intersections only deduplicate types with the same object identity.

**Comment by fatcerberus** [comment]
Ah, thanks. I had assumed that type aliases worked like macros and were always expanded in-place.

**Comment by schani** [comment]
Thank you so much for the workaround!

**Comment by amcasey** [comment]
Confirmed that the mitigation works in my unreduced original repro.

@schani How does the editor feel in the real-world usage?

**Comment by schani** [comment]
In the small repo linked to above it's super fast now.  In our big proprietary repo performance is definitely improved, but still sluggish.  This was just one source of slowdown.

**Comment by Feirell** [comment]
If you want to have a isolated example, with the generated trace files, have a look at an identical issue: https://github.com/microsoft/TypeScript/issues/46840 . It is somewhat more bloated but has the same underlying issue and the solution from @ahejlsberg works and brings the check time from 50s down to 0.25s.

Thank you very much!

**Comment by rotu** [comment]
This seems to have been fixed somewhere between 5.0.4 and 5.1.6
The intersection still grows with each chained call, but inference no longer fails with "Type instantiation is excessively deep and possibly infinite."
